src/linalg.jl

```diff
@@ -149,3 +149,21 @@ function _mul!(C::BlockDiagonal, A::BlockDiagonal, B::BlockDiagonal, α::Number,
 
     return C
 end
+
+function LinearAlgebra.:\(B::BlockDiagonal, vm::AbstractVecOrMat)
+    row_i = 1
+    # BlockDiagonals with non-square blocks
+    all(b -> size(b, 1) == size(b, 2), blocks(B)) || throw(
+        ArgumentError(
+            "Left division (\\) is not defined for BlockDiagonals with " *
+            "non-square blocks (they are singular matrices).",
+        ),
+    )
+    result = similar(vm)
+    for block in blocks(B)
+        nrow = size(block, 1)
+        result[row_i:(row_i + nrow - 1), :] = block \ vm[row_i:(row_i + nrow - 1), :]
```

Comment:
i've not thought too much, but perhaps you know: is there any rewriting of this that would allow us to access `vm` col-wise rather than row-wise?

also does using views of `vm` help performance here at all?

Reply:
I don't think accessing `vm` colwise is possible for this problem.
`views` could help I suppose. I am going to run some benchmark and see

Reply:
Views does not seem to help that much in this simple example
```julia
julia> using BenchmarkTools
julia> N = 400
julia> A = BlockDiagonal([rand(N, N) for _ in 1:5])
julia> x = rand(5N, 100)
julia> A \ x
julia> @btime $A \ $x # No views
  7.085 ms (37 allocations: 10.70 MiB)
julia> @btime $A \ $x # With views
  7.403 ms (27 allocations: 9.17 MiB)
```

Reply:
i think the reduced allocations suggests let's use views:

e.g. on a matrix of the same size, but more blocks:
```jl
julia> M = 40
julia> B = BlockDiagonal([rand(M, M) for _ in 1:5])
julia> x = rand(5N, 100)
julia> @btime $A \ $x # No views
  2.155 ms (302 allocations: 5.22 MiB)
julia> @btime $A \ $x # With views
  2.264 ms (202 allocations: 3.69 MiB)
```